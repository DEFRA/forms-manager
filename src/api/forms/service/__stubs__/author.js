export default {
  id: 'f50ceeed-b7a4-47cf-a498-094efc99f8bc',
  displayName: 'Enrique Chase'
}
