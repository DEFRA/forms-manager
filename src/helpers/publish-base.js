import { PublishCommand } from '@aws-sdk/client-sns'

import { config } from '~/src/config/index.js'
import { getSNSClient } from '~/src/helpers/sns.js'

const snsTopicArn = config.get('snsTopicArn')

const client = getSNSClient()

/**
 * Publish event onto topic
 * @param {Message} message
 */
export function publishEvent(message) {
  const command = new PublishCommand({
    TopicArn: snsTopicArn,
    Message: JSON.stringify(message)
  })

  return client.send(command)
}

/**
 * @import { FormMetadata, Message, FormCreatedMessage, FormCreatedMessageData } from '@defra/forms-model'
 */
