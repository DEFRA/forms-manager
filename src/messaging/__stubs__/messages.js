import {
  AuditEventMessageCategory,
  AuditEventMessageSchemaVersion,
  AuditEventMessageSource,
  AuditEventMessageType
} from '@defra/forms-model'

import createdBy from '~/src/api/forms/service/__stubs__/author.js'

const createdAt = new Date('2025-07-26')
const schemaVersion = AuditEventMessageSchemaVersion.V1
const messageCreatedAt = createdAt
const formId = '1e0bc245-1577-431a-9026-1db38af821c6'
const entityId = formId
const slug = 'audit-form'
const dataBase = {
  formId,
  slug
}

/**
 * @param {Partial<MessageBase>} partialFormMessageBase
 * @returns {Omit<MessageBase, 'type'|'category'|'entityId'|'data'>}
 */
export function buildFormMessageBase(partialFormMessageBase = {}) {
  return {
    schemaVersion,
    source: AuditEventMessageSource.FORMS_MANAGER,
    messageCreatedAt,
    createdAt,
    createdBy,
    ...partialFormMessageBase
  }
}

/**
 * @param {Partial<FormTitleUpdatedMessage>} partialFormTitleUpdatedMessage
 * @returns {FormTitleUpdatedMessage}
 */
export function buildFormTitleUpdatedMessage(
  partialFormTitleUpdatedMessage = {}
) {
  return {
    ...buildFormMessageBase(),
    entityId,
    source: AuditEventMessageSource.FORMS_MANAGER,
    schemaVersion: AuditEventMessageSchemaVersion.V1,
    category: AuditEventMessageCategory.FORM,
    type: AuditEventMessageType.FORM_TITLE_UPDATED,
    data: {
      ...dataBase,
      changes: {
        previous: {
          title: 'Previous title'
        },
        new: {
          title: 'New title'
        }
      }
    },
    ...partialFormTitleUpdatedMessage
  }
}

/**
 * @param {Partial<FormOrganisationUpdatedMessage>} partialFormOrganisationUpdatedMessage
 * @returns {FormOrganisationUpdatedMessage}
 */
export function buildFormOrganisationUpdatedMessage(
  partialFormOrganisationUpdatedMessage = {}
) {
  return {
    ...buildFormMessageBase(),
    entityId,
    source: AuditEventMessageSource.FORMS_MANAGER,
    schemaVersion: AuditEventMessageSchemaVersion.V1,
    category: AuditEventMessageCategory.FORM,
    type: AuditEventMessageType.FORM_ORGANISATION_UPDATED,
    data: {
      ...dataBase,
      changes: {
        previous: {
          organisation: 'Natural England'
        },
        new: {
          organisation: 'Defra'
        }
      }
    },
    ...partialFormOrganisationUpdatedMessage
  }
}

/**
 * @import { MessageBase, AuditMessage, FormTitleUpdatedMessage, FormOrganisationUpdatedMessage } from '@defra/forms-model'
 */
