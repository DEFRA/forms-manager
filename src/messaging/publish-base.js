import { PublishCommand } from '@aws-sdk/client-sns'

import { logger } from '~/src/api/forms/service/shared.js'
import { config } from '~/src/config/index.js'
import { getSNSClient } from '~/src/messaging/sns.js'

const snsTopicArn = config.get('snsTopicArn')

const client = getSNSClient()

/**
 * Publish event onto topic
 * @param {AuditMessage} message
 */
export async function publishEvent(message) {
  const shouldPublish = config.get('publishAuditEvents')

  if (shouldPublish) {
    const command = new PublishCommand({
      TopicArn: snsTopicArn,
      Message: JSON.stringify(message)
    })

    const result = await client.send(command)

    logger.info(
      `Published ${message.type} event for formId ${message.entityId}. MessageId: ${result.MessageId}`
    )

    return result
  }

  return undefined
}

/**
 * @import { AuditMessage } from '@defra/forms-model'
 */
