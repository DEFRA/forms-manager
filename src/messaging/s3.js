/**
 * @param {FormDefinition} _formDefinition
 * @returns {Promise<{
 *   slug?: string;
 *   fileId?: string;
 *   filename?: string;
 * }>}
 */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function saveToS3(_formDefinition) {
  // TODO: implement saveToS3 function
  return Promise.resolve({})
}

/**
 * @import { FormDefinition } from '@defra/forms-model'
 */
