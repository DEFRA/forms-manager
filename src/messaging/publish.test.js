import {
  AuditEventMessageCategory,
  AuditEventMessageSchemaVersion,
  AuditEventMessageSource,
  AuditEventMessageType,
  FormDefinitionRequestType
} from '@defra/forms-model'
import {
  buildDefinition,
  buildMetaData,
  buildQuestionPage
} from '@defra/forms-model/stubs'
import { ValidationError } from 'joi'

import author from '~/src/api/forms/service/__stubs__/author.js'
import {
  BASE_CREATED_DATE,
  formMetadataDocument
} from '~/src/api/forms/service/__stubs__/service.js'
import { buildFormOrganisationUpdatedMessage } from '~/src/messaging/__stubs__/messages.js'
import { publishEvent } from '~/src/messaging/publish-base.js'
import {
  bulkPublishEvents,
  publishDraftCreatedFromLiveEvent,
  publishFormCreatedEvent,
  publishFormDraftDeletedEvent,
  publishFormDraftReplacedEvent,
  publishFormMigratedEvent,
  publishFormTitleUpdatedEvent,
  publishFormUpdatedEvent,
  publishLiveCreatedFromDraftEvent
} from '~/src/messaging/publish.js'
import { saveToS3 } from '~/src/messaging/s3.js'
jest.mock('~/src/messaging/s3.js')

jest.mock('~/src/messaging/publish-base.js')

describe('publish', () => {
  const formId = '689b7ab1d0eeac9711a7fb33'
  const slug = 'audit-form'
  const title = 'My Audit Event Form'
  const organisation = 'Defra'
  const teamName = 'Forms'
  const teamEmail = 'forms@example.uk'
  const createdAt = new Date('2025-07-23')
  const createdBy = {
    id: '83f09a7d-c80c-4e15-bcf3-641559c7b8a7',
    displayName: 'Enrique Chase'
  }
  const updatedAt = new Date('2025-07-24')
  const updatedBy = {
    id: 'a53b4360-bdf6-4d13-8975-25032ce76312',
    displayName: 'Gandalf'
  }
  const messageId = '2888a402-7609-43c5-975f-b1974969cdb6'
  const metadata = buildMetaData({
    id: formId,
    slug,
    title,
    organisation,
    teamName,
    teamEmail,
    createdAt,
    createdBy,
    updatedAt,
    updatedBy
  })

  beforeEach(() => {
    jest.mocked(publishEvent).mockResolvedValue({
      MessageId: '2888a402-7609-43c5-975f-b1974969cdb6',
      SequenceNumber: undefined,
      $metadata: {}
    })
  })
  afterEach(() => {
    jest.resetAllMocks()
  })

  describe('publishFormCreatedEvent', () => {
    it('should publish FORM_CREATED event', async () => {
      await publishFormCreatedEvent(metadata)

      expect(publishEvent).toHaveBeenCalledWith({
        entityId: formId,
        source: AuditEventMessageSource.FORMS_MANAGER,
        messageCreatedAt: expect.any(Date),
        schemaVersion: AuditEventMessageSchemaVersion.V1,
        category: AuditEventMessageCategory.FORM,
        type: AuditEventMessageType.FORM_CREATED,
        createdAt,
        createdBy,
        data: {
          formId,
          slug,
          title,
          organisation,
          teamName,
          teamEmail
        }
      })
      const [publishEventCall] = jest.mocked(publishEvent).mock.calls[0]
      expect(publishEventCall).toMatchSnapshot({
        messageCreatedAt: expect.any(Date)
      })
    })

    it('should not publish the event if the schema is incorrect', async () => {
      jest.mocked(publishEvent).mockRejectedValue(new Error('rejected'))
      const invalidMetaData = {
        title,
        organisation,
        teamName,
        teamEmail,
        createdAt,
        createdBy
      }

      // @ts-expect-error - invalid schema
      await expect(publishFormCreatedEvent(invalidMetaData)).rejects.toThrow(
        new ValidationError(
          '"entityId" is required. "data.formId" is required. "data.slug" is required',
          [],
          invalidMetaData
        )
      )
    })
  })

  describe('publishFormTitleUpdatedEvent', () => {
    it('should publish FORM_TITLE_UPDATED event', async () => {
      const oldMetadata = buildMetaData({
        ...metadata,
        title: 'Old form title'
      })
      const response = await publishFormTitleUpdatedEvent(metadata, oldMetadata)
      expect(response?.MessageId).toBe(messageId)
      expect(publishEvent).toHaveBeenCalledWith({
        entityId: formId,
        messageCreatedAt: expect.any(Date),
        source: AuditEventMessageSource.FORMS_MANAGER,
        schemaVersion: AuditEventMessageSchemaVersion.V1,
        category: AuditEventMessageCategory.FORM,
        type: AuditEventMessageType.FORM_TITLE_UPDATED,
        createdAt: updatedAt,
        createdBy: updatedBy,
        data: {
          formId,
          slug: 'audit-form',
          changes: {
            previous: {
              title: 'Old form title'
            },
            new: {
              title
            }
          }
        }
      })
      const [publishEventCall] = jest.mocked(publishEvent).mock.calls[0]
      expect(publishEventCall).toMatchSnapshot({
        messageCreatedAt: expect.any(Date)
      })
    })
  })

  describe('publishLiveCreatedFromDraftEvent', () => {
    it('should publish FORM_LIVE_CREATED_FROM_DRAFT event', async () => {
      const response = await publishLiveCreatedFromDraftEvent(
        formId,
        updatedAt,
        updatedBy
      )

      expect(response?.MessageId).toBe(messageId)
      expect(publishEvent).toHaveBeenCalledWith({
        entityId: formId,
        source: AuditEventMessageSource.FORMS_MANAGER,
        messageCreatedAt: expect.any(Date),
        schemaVersion: AuditEventMessageSchemaVersion.V1,
        category: AuditEventMessageCategory.FORM,
        type: AuditEventMessageType.FORM_LIVE_CREATED_FROM_DRAFT,
        createdAt: updatedAt,
        createdBy: updatedBy
      })
      const [publishEventCall] = jest.mocked(publishEvent).mock.calls[0]
      expect(publishEventCall).toMatchSnapshot({
        messageCreatedAt: expect.any(Date)
      })
    })
  })

  describe('publishDraftCreatedFromLiveEvent', () => {
    it('should publish FORM_DRAFT_CREATED_FROM_LIVE event', async () => {
      const response = await publishDraftCreatedFromLiveEvent(
        formId,
        updatedAt,
        updatedBy
      )

      expect(response?.MessageId).toBe(messageId)
      expect(publishEvent).toHaveBeenCalledWith({
        entityId: formId,
        source: AuditEventMessageSource.FORMS_MANAGER,
        messageCreatedAt: expect.any(Date),
        schemaVersion: AuditEventMessageSchemaVersion.V1,
        category: AuditEventMessageCategory.FORM,
        type: AuditEventMessageType.FORM_DRAFT_CREATED_FROM_LIVE,
        createdAt: updatedAt,
        createdBy: updatedBy
      })
      const [publishEventCall] = jest.mocked(publishEvent).mock.calls[0]
      expect(publishEventCall).toMatchSnapshot({
        messageCreatedAt: expect.any(Date)
      })
    })
  })

  describe('publishFormDraftDeletedEvent', () => {
    it('should publish a FORM_DRAFT_DELETED event', async () => {
      const response = await publishFormDraftDeletedEvent(metadata, author)
      expect(response?.MessageId).toBe(messageId)
      expect(publishEvent).toHaveBeenCalledWith({
        entityId: formId,
        source: AuditEventMessageSource.FORMS_MANAGER,
        messageCreatedAt: expect.any(Date),
        schemaVersion: AuditEventMessageSchemaVersion.V1,
        category: AuditEventMessageCategory.FORM,
        type: AuditEventMessageType.FORM_DRAFT_DELETED,
        createdAt: expect.any(Date),
        createdBy: author,
        data: {
          formId,
          slug
        }
      })
      const [publishEventCall] = jest.mocked(publishEvent).mock.calls[0]
      expect(publishEventCall).toMatchSnapshot({
        messageCreatedAt: expect.any(Date),
        createdAt: expect.any(Date)
      })
    })
  })

  describe('publishFormMigratedEvent', () => {
    it('should publish FORM_MIGRATED event', async () => {
      const response = await publishFormMigratedEvent(
        formId,
        updatedAt,
        updatedBy
      )

      expect(response?.MessageId).toBe(messageId)
      expect(publishEvent).toHaveBeenCalledWith({
        entityId: formId,
        source: AuditEventMessageSource.FORMS_MANAGER,
        messageCreatedAt: expect.any(Date),
        schemaVersion: AuditEventMessageSchemaVersion.V1,
        category: AuditEventMessageCategory.FORM,
        type: AuditEventMessageType.FORM_MIGRATED,
        createdAt: updatedAt,
        createdBy: updatedBy
      })
      const [publishEventCall] = jest.mocked(publishEvent).mock.calls[0]
      expect(publishEventCall).toMatchSnapshot({
        messageCreatedAt: expect.any(Date)
      })
    })
  })

  describe('publishFormUpdatedEvent', () => {
    it('should publish a FORM_UPDATED event', async () => {
      const requestType = FormDefinitionRequestType.CREATE_PAGE
      const payload = buildQuestionPage({})
      await publishFormUpdatedEvent(formMetadataDocument, payload, requestType)

      const [publishEventCall] = jest.mocked(publishEvent).mock.calls[0]
      expect(publishEventCall).toMatchObject({
        schemaVersion: AuditEventMessageSchemaVersion.V1,
        category: AuditEventMessageCategory.FORM,
        type: AuditEventMessageType.FORM_UPDATED,
        createdAt: BASE_CREATED_DATE,
        createdBy: author
      })
      expect(publishEventCall.data).toMatchObject({
        requestType,
        s3Meta: undefined,
        slug: formMetadataDocument.slug,
        payload
      })
      expect(publishEventCall).toMatchSnapshot({
        messageCreatedAt: expect.any(Date)
      })
    })
  })

  describe('publishFormDraftReplacedEvent', () => {
    it('should publish a REPLACE_DRAFT event', async () => {
      const s3Meta = {
        fileId: '3HL4kqtJlcpXrof3W3Zz4YBdvdz2FJ9n',
        filename: '6883d8667a2a64da10af4312.json',
        s3Key: 'audit-definitions/6883d8667a2a64da10af4312.json'
      }
      jest.mocked(saveToS3).mockResolvedValue(s3Meta)
      const requestType = FormDefinitionRequestType.REPLACE_DRAFT
      const definition = buildDefinition()
      await publishFormDraftReplacedEvent(formMetadataDocument, definition)

      const [publishEventCall] = jest.mocked(publishEvent).mock.calls[0]
      expect(publishEventCall).toMatchObject({
        schemaVersion: AuditEventMessageSchemaVersion.V1,
        category: AuditEventMessageCategory.FORM,
        type: AuditEventMessageType.FORM_UPDATED,
        createdAt: BASE_CREATED_DATE,
        createdBy: author
      })
      expect(publishEventCall.data).toMatchObject({
        requestType,
        s3Meta,
        slug: formMetadataDocument.slug,
        payload: undefined
      })
      expect(publishEventCall).toMatchSnapshot({
        messageCreatedAt: expect.any(Date)
      })
    })
  })

  describe('bulkPublishEvents', () => {
    it('should publish FORM_ORGANISATION_UPDATED', async () => {
      const message = buildFormOrganisationUpdatedMessage()
      const result = await bulkPublishEvents([message])
      expect(result).toEqual([
        {
          type: AuditEventMessageType.FORM_ORGANISATION_UPDATED,
          messageId: '2888a402-7609-43c5-975f-b1974969cdb6'
        }
      ])
      expect(publishEvent).toHaveBeenCalledWith(message)
      const [publishEventCall] = jest.mocked(publishEvent).mock.calls[0]
      expect(publishEventCall).toMatchSnapshot({
        messageCreatedAt: expect.any(Date)
      })
    })

    it('should fail given rejection', async () => {
      jest.mocked(publishEvent).mockRejectedValue(new Error('an error'))
      const message = buildFormOrganisationUpdatedMessage()

      await expect(bulkPublishEvents([message])).rejects.toThrow()
    })
  })
})
