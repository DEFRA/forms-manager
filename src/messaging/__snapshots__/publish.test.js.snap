// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`publish bulkPublishEvents should publish FORM_ORGANISATION_UPDATED 1`] = `
{
  "category": "FORM",
  "createdAt": 2025-07-26T00:00:00.000Z,
  "createdBy": {
    "displayName": "Enrique Chase",
    "id": "f50ceeed-b7a4-47cf-a498-094efc99f8bc",
  },
  "data": {
    "changes": {
      "new": {
        "organisation": "Defra",
      },
      "previous": {
        "organisation": "Natural England",
      },
    },
    "formId": "689b7ab1d0eeac9711a7fb33",
    "slug": "audit-form",
  },
  "entityId": "689b7ab1d0eeac9711a7fb33",
  "messageCreatedAt": Any<Date>,
  "schemaVersion": 1,
  "source": "FORMS_MANAGER",
  "type": "FORM_ORGANISATION_UPDATED",
}
`;

exports[`publish publishDraftCreatedFromLiveEvent should publish FORM_DRAFT_CREATED_FROM_LIVE event 1`] = `
{
  "category": "FORM",
  "createdAt": 2025-07-24T00:00:00.000Z,
  "createdBy": {
    "displayName": "Gandalf",
    "id": "a53b4360-bdf6-4d13-8975-25032ce76312",
  },
  "entityId": "689b7ab1d0eeac9711a7fb33",
  "messageCreatedAt": Any<Date>,
  "schemaVersion": 1,
  "source": "FORMS_MANAGER",
  "type": "FORM_DRAFT_CREATED_FROM_LIVE",
}
`;

exports[`publish publishFormCreatedEvent should publish FORM_CREATED event 1`] = `
{
  "category": "FORM",
  "createdAt": 2025-07-23T00:00:00.000Z,
  "createdBy": {
    "displayName": "Enrique Chase",
    "id": "83f09a7d-c80c-4e15-bcf3-641559c7b8a7",
  },
  "data": {
    "formId": "689b7ab1d0eeac9711a7fb33",
    "organisation": "Defra",
    "slug": "audit-form",
    "teamEmail": "forms@example.uk",
    "teamName": "Forms",
    "title": "My Audit Event Form",
  },
  "entityId": "689b7ab1d0eeac9711a7fb33",
  "messageCreatedAt": Any<Date>,
  "schemaVersion": 1,
  "source": "FORMS_MANAGER",
  "type": "FORM_CREATED",
}
`;

exports[`publish publishFormDraftDeletedEvent should publish a FORM_DRAFT_DELETED event 1`] = `
{
  "category": "FORM",
  "createdAt": Any<Date>,
  "createdBy": {
    "displayName": "Enrique Chase",
    "id": "f50ceeed-b7a4-47cf-a498-094efc99f8bc",
  },
  "data": {
    "formId": "689b7ab1d0eeac9711a7fb33",
    "slug": "audit-form",
  },
  "entityId": "689b7ab1d0eeac9711a7fb33",
  "messageCreatedAt": Any<Date>,
  "schemaVersion": 1,
  "source": "FORMS_MANAGER",
  "type": "FORM_DRAFT_DELETED",
}
`;

exports[`publish publishFormDraftReplacedEvent should publish a REPLACE_DRAFT event 1`] = `
{
  "category": "FORM",
  "createdAt": 2020-01-01T00:00:00.000Z,
  "createdBy": {
    "displayName": "Enrique Chase",
    "id": "f50ceeed-b7a4-47cf-a498-094efc99f8bc",
  },
  "data": {
    "formId": "661e4ca5039739ef2902b214",
    "payload": undefined,
    "requestType": "REPLACE_DRAFT",
    "s3Meta": {
      "fileId": "3HL4kqtJlcpXrof3W3Zz4YBdvdz2FJ9n",
      "filename": "6883d8667a2a64da10af4312.json",
      "s3Key": "audit-definitions/6883d8667a2a64da10af4312.json",
    },
    "slug": "test-form",
  },
  "entityId": "661e4ca5039739ef2902b214",
  "messageCreatedAt": Any<Date>,
  "schemaVersion": 1,
  "source": "FORMS_MANAGER",
  "type": "FORM_UPDATED",
}
`;

exports[`publish publishFormMigratedEvent should publish FORM_MIGRATED event 1`] = `
{
  "category": "FORM",
  "createdAt": 2025-07-24T00:00:00.000Z,
  "createdBy": {
    "displayName": "Gandalf",
    "id": "a53b4360-bdf6-4d13-8975-25032ce76312",
  },
  "entityId": "689b7ab1d0eeac9711a7fb33",
  "messageCreatedAt": Any<Date>,
  "schemaVersion": 1,
  "source": "FORMS_MANAGER",
  "type": "FORM_MIGRATED",
}
`;

exports[`publish publishFormTitleUpdatedEvent should publish FORM_TITLE_UPDATED event 1`] = `
{
  "category": "FORM",
  "createdAt": 2025-07-24T00:00:00.000Z,
  "createdBy": {
    "displayName": "Gandalf",
    "id": "a53b4360-bdf6-4d13-8975-25032ce76312",
  },
  "data": {
    "changes": {
      "new": {
        "title": "My Audit Event Form",
      },
      "previous": {
        "title": "Old form title",
      },
    },
    "formId": "689b7ab1d0eeac9711a7fb33",
    "slug": "audit-form",
  },
  "entityId": "689b7ab1d0eeac9711a7fb33",
  "messageCreatedAt": Any<Date>,
  "schemaVersion": 1,
  "source": "FORMS_MANAGER",
  "type": "FORM_TITLE_UPDATED",
}
`;

exports[`publish publishFormUpdatedEvent should publish a FORM_UPDATED event 1`] = `
{
  "category": "FORM",
  "createdAt": 2020-01-01T00:00:00.000Z,
  "createdBy": {
    "displayName": "Enrique Chase",
    "id": "f50ceeed-b7a4-47cf-a498-094efc99f8bc",
  },
  "data": {
    "formId": "661e4ca5039739ef2902b214",
    "payload": {
      "components": [],
      "id": "ffefd409-f3f4-49fe-882e-6e89f44631b1",
      "next": [],
      "path": "/page-one",
      "title": "Page One",
    },
    "requestType": "CREATE_PAGE",
    "s3Meta": undefined,
    "slug": "test-form",
  },
  "entityId": "661e4ca5039739ef2902b214",
  "messageCreatedAt": Any<Date>,
  "schemaVersion": 1,
  "source": "FORMS_MANAGER",
  "type": "FORM_UPDATED",
}
`;

exports[`publish publishLiveCreatedFromDraftEvent should publish FORM_LIVE_CREATED_FROM_DRAFT event 1`] = `
{
  "category": "FORM",
  "createdAt": 2025-07-24T00:00:00.000Z,
  "createdBy": {
    "displayName": "Gandalf",
    "id": "a53b4360-bdf6-4d13-8975-25032ce76312",
  },
  "entityId": "689b7ab1d0eeac9711a7fb33",
  "messageCreatedAt": Any<Date>,
  "schemaVersion": 1,
  "source": "FORMS_MANAGER",
  "type": "FORM_LIVE_CREATED_FROM_DRAFT",
}
`;
